REGRESSION: "Use Apple Vision Pro" button disappears on FX Alien Earth Website in Safari after entering Spatial viewer and exiting

vitorroriz opened the pull request:
#### 33ebcf97f481d763ede9c79c154235240f291c77
<pre>
REGRESSION: &quot;Use Apple Vision Pro&quot; button disappears on FX Alien Earth Website in Safari after entering Spatial viewer and exiting
<a href="https://bugs.webkit.org/show_bug.cgi?id=299646">https://bugs.webkit.org/show_bug.cgi?id=299646</a>
<a href="https://rdar.apple.com/158351089">rdar://158351089</a>

Reviewed by Simon Fraser.

WebFullScreenManager::didExitFullScreen issues a scrollIntoViewIfNotVisible.
This allows for scrolling into an an area that is unscrollable for the user.
We are adding a new option (ShouldAvoidScrollingToUnscrollableArea) to
ScrollRectToVisibleOptions such that ::didExitFullScreen can opt-out for
this behavior.

Tests: imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-hidden.html
       imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden-ref.html
       imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden.html
       imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-ref.html
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-hidden-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-hidden.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-ref.html: Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::scrollIntoViewIfNotVisible):
* Source/WebCore/dom/Element.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFocusedElementInternal):
(WebCore::adjustScrollRectToVisibleOptionsForHiddenOverflow):
(WebCore::LocalFrameView::scrollRectToVisible):
* Source/WebCore/platform/ScrollAlignment.cpp:
* Source/WebCore/platform/ScrollAlignment.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::shouldTryToScrollForScrollIntoView const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::didExitFullScreen):

Canonical link: <a href="https://commits.webkit.org/300851@main">https://commits.webkit.org/300851@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8abe960ddff28f56494a22fb201f23fb5a918fa1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124081 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/43784 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34492 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130910 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125958 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44518 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52383 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94388 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127035 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/35477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110998 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74980 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29162 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74394 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/105215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133582 "Built successfully") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/51022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38874 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102855 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51399 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107217 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102666 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26113 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/48023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/26273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/47900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/50877 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56648 "Built successfully") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50340 "Build is in progress. Recent messages:") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/53687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/52014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/dom/Element.h:
```diff
@@ -112,6 +112,7 @@ enum class ShadowRootDelegatesFocus : bool { No, Yes };
 enum class ShadowRootMode : uint8_t;
 enum class ShadowRootClonable : bool { No, Yes };
 enum class ShadowRootSerializable : bool { No, Yes };
+enum class ShouldAvoidScrollingToUnscrollableArea : bool { No, Yes };
```
I would call this `AllowScrollingOverflowHidden`

smfr commented on Source/WebCore/dom/Element.h:
```diff
@@ -299,7 +300,8 @@ class Element : public ContainerNode {
     WEBCORE_EXPORT void scrollIntoViewIfNeeded(bool centerIfNeeded = true);
-    WEBCORE_EXPORT void scrollIntoViewIfNotVisible(bool centerIfNotVisible = true);
+    WEBCORE_EXPORT void scrollIntoViewIfNotVisible(bool centerIfNotVisible = true, ShouldAvoidScrollingToUnscrollableArea = ShouldAvoidScrollingToUnscrollableArea::No);
+    WEBCORE_EXPORT void scrollIntoViewIfScrollableAndNotVisible(bool centerIfNotVisible = true);
```
Is there any utility in having `scrollIntoViewIfScrollableAndNotVisible`?

vitorroriz replied:
I just wanted to abstract the "scrollable" variant on its own function, but I can drop it

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3220,7 +3220,7 @@ void LocalFrameView::scrollToFocusedElementInternal()
     auto anchorRectWithScrollMargin = absoluteBounds.marginRect;
     auto anchorRect = absoluteBounds.anchorRect;
-    LocalFrameView::scrollRectToVisible(anchorRectWithScrollMargin, *renderer, insideFixed, { m_selectionRevealModeForFocusedElement, ScrollAlignment::alignCenterIfNeeded, ScrollAlignment::alignCenterIfNeeded, ShouldAllowCrossOriginScrolling::No, ScrollBehavior::Auto, OnlyAllowForwardScrolling::No, anchorRect });
+    LocalFrameView::scrollRectToVisible(anchorRectWithScrollMargin, *renderer, insideFixed, { m_selectionRevealModeForFocusedElement, ScrollAlignment::alignCenterIfNeeded, ScrollAlignment::alignCenterIfNeeded, ShouldAllowCrossOriginScrolling::No, ScrollBehavior::Auto, OnlyAllowForwardScrolling::No, ShouldAvoidScrollingToUnscrollableArea::No, anchorRect });
```
This is a behavior change for `focus()`, right? This deserves a (non-fullscreen) test. What do other browsers do?

vitorroriz replied:
This does not intend to change behavior regarding focus, I just needed to pass the default value of `ShouldAvoidScrollingToUnscrollableArea` due to the way options are being initialized here. I can replace it by a initialization list with named members instead of relying in order if you think it is better.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3272,6 +3278,8 @@ bool LocalFrameView::scrollRectToVisible(const LayoutRect& absoluteRect, const R
 
     for (; layer; layer = layer->enclosingContainingBlockLayer(CrossFrameBoundaries::No)) {
         if (layer->shouldTryToScrollForScrollIntoView()) {
+            if (shouldAvoidScrollingToUnscrollableArea(adjustedOptions, layer->renderer().style()))
```
I think this would go better inside `shouldTryToScrollForScrollIntoView()`, passing `adjustedOptions`.

smfr commented on LayoutTests/imported/w3c/web-platform-tests/fullscreen/model/exit-fullscreen-scroll-to-unscrollable-area.html:
```diff
@@ -0,0 +1,38 @@
+<!DOCTYPE html>
```
Is fullscreen interoperable enough to have a WPT?

smfr commented on LayoutTests/imported/w3c/web-platform-tests/fullscreen/model/exit-fullscreen-scroll-to-unscrollable-area-expected.html:
```diff
@@ -0,0 +1,15 @@
+<body>
+<div class="container">
+  <div style="width: 2500px; height: 500px; background-color: green;"></div>
+  <button class="button-style" type="button" id="vision-pro-button"> Request fullscreen!</button>
```
Don't use something like "vision-pro-button" in a WPT. Make it more generic.

vitorroriz replied:
oops, I've forgot to clean this up for the -expected/ref. Thanks.

smfr commented on LayoutTests/imported/w3c/web-platform-tests/fullscreen/model/exit-fullscreen-scroll-to-unscrollable-area.html:
```diff
@@ -0,0 +1,38 @@
+</body>
+</html>
+<script>
+  if (window.testRunner)
```
You can't use `window.testRunner` in a WPT. You have to use the WPT harness.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3254,6 +3254,12 @@ void LocalFrameView::cancelScheduledTextFragmentIndicatorTimer()
+static bool shouldAvoidScrollingToUnscrollableArea(const ScrollRectToVisibleOptions& options, const RenderStyle& style)
+{
+    // Can we check overflowX || overflowY or should know in what what axis its actually overflowing?
+    return options.shouldAvoidScrollingToUnscrollableArea == ShouldAvoidScrollingToUnscrollableArea::Yes && (style.overflowX() == Overflow::Hidden || style.overflowY() == Overflow::Hidden);
```
Reading the code in `Adjuster::adjust()` I think we can have `hidden` on one axis and `auto` or 'scroll` on the other, so this flag needs to prevent scrolling per axis.

vitorroriz committed:
REGRESSION: "Use Apple Vision Pro" button disappears on FX Alien Earth Website in Safari after entering Spatial viewer and exiting
https://bugs.webkit.org/show_bug.cgi?id=299646
rdar://158351089

Reviewed by Simon Fraser.

WebFullScreenManager::didExitFullScreen issues a scrollIntoViewIfNotVisible.
This allows for scrolling into an an area that is unscrollable for the user.
We are adding a new option (ShouldAvoidScrollingToUnscrollableArea) to
ScrollRectToVisibleOptions such that ::didExitFullScreen can opt-out for
this behavior.

Tests: imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-hidden.html
       imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden-ref.html
       imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden.html
       imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-ref.html
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-hidden-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-hidden.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-overflow-x-hidden.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/exit-fullscreen-scroll-to-unscrollable-area-ref.html: Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::scrollIntoViewIfNotVisible):
* Source/WebCore/dom/Element.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFocusedElementInternal):
(WebCore::adjustScrollRectToVisibleOptionsForHiddenOverflow):
(WebCore::LocalFrameView::scrollRectToVisible):
* Source/WebCore/platform/ScrollAlignment.cpp:
* Source/WebCore/platform/ScrollAlignment.h:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::shouldTryToScrollForScrollIntoView const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::didExitFullScreen):

Canonical link: https://commits.webkit.org/300851@main
